src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc

```diff
@@ -73,16 +73,71 @@ void ApplyPerturbation(XvectorPerturbOptions opts,
   }
   // Perturb speed of signal egs
   Matrix<BaseFloat> warped_distorted_shifted_egs(rand_distort_shifted_egs);
-  if (opts.max_time_stretch != 0.0) 
-    perturb_xvector.TimeStretch(rand_distort_shifted_egs, 
-                                &warped_distorted_shifted_egs);
+  ////if (opts.max_time_stretch != 0.0) 
+  ////  perturb_xvector.TimeStretch(rand_distort_shifted_egs, 
+  ////                              &warped_distorted_shifted_egs);
    
   // If nagation is true, the sample values are randomly negated
   // with some probability.
-  if (opts.negation) {
+  ////if (opts.negation) {
    
+  ////}
+}
+
+// add
+// This function add the noise to the orginial signal. We should not normalize 
+// the signal level of the orginial signal. According to SNR, we rescale the noise
+// and add it. So that the perturbed signal is created. 
+void ApplyAddAdditiveNoise(const int32 &SNR,
+                           const Matrix<BaseFloat> &input_eg,
+                           const Matrix<BaseFloat> &noise_eg,
+                           Matrix<BaseFloat> *perturb_eg) {
+  // In the version, we ask the noise_cols >= input_cols. If mfcc, the cols are equal.
+  // If raw data, we ask the noise_cols > input_cols.
+  int32 input_rows = input_eg.NumRows(), input_cols = input_eg.NumCols();  
+  KALDI_ASSERT(noise_eg.NumCols() >= input_cols);
```

Comment:
For now, we can focus on MFCC domain, and if it gives us improvement, we can switch to raw waveform.
We may need to write different Perturbation class for raw waveform as we have more flexibility in raw waveform domain.